[webpack] Introduce webpack-embedded-dependencies-plugin (#3975)
microsoft/rushstack: webpack/webpack-embedded-dependencies-plugin/README.md

# webpack-embedded-dependencies-plugin

## Installation

`npm install @rushstack/webpack-embedded-dependencies-plugin --save`

## Overview
A webpack plugin for generating a list of embedded dependencies. Embedded dependencies are third-party packages which are being

> bundled into your released code and are often times subject to license, security, and other legal requirements. This plugin

Reviewer: ```suggestion
bundled into your released code and are oftentimes subject to license, security, and other legal requirements. This plugin
```

aims to make it easier to generate a list of embedded dependencies and their associated metadata, so they can be analyzed by additional tools.

The plugin also includes the ability to generate a secondary asset which contains the license text for each embedded dependency into a single file called

> THIRD-PARTY-NOTICES.html. This is a common legal requirement for large companies deploying commercial services/products containing open source code.

Reviewer: ```suggestion
THIRD-PARTY-NOTICES.html. This is a common legal requirement when deploying services or products containing open-source code.
```

> THIRD-PARTY-NOTICES.html. This is a common legal requirement for large companies deploying commercial services/products containing open source code.

Reviewer: HTML? Seems like the name and output format should be configurable.

Reviewer: Oh, the name is configurable. This summary seems to imply that it isn't.

HTML seems like a weird default choice.

Author: When reviewing existing implementations, each major product was generating an HTML file. Since this is an optional generation, having the most common scenario covered felt right to me.

      .map((pkg) => {
        return `<h2>${pkg.name}</h2><p>${pkg.license}</p>`;
      }).join('');
  }
})
```

### `generatedLicenseFilename`: `LicenseFileName`

```typescript
new EmbeddedDependenciesWebpackPlugin({
  generateLicenseFile: true,
  generatedLicenseFilename: 'custom-license-file-name.html'
})
```

Name of the generated license file. Defaults to THIRD-PARTY-NOTICES.html

## Types
<!--
/**
 * @alpha
 * Function type that generates the license file.
 *
 * @example
 * ```ts
 * const licenseFileGenerator: LicenseFileGeneratorFunction = (packages: IPackageData[]): string => {
 *  return packages
 *   .map((pkg) => {
 *    return `<h2>${pkg.name}</h2><p>${pkg.license}</p>`;
 *  }).join('');
 * }
 * ```
 */
export declare type LicenseFileGeneratorFunction = (packages: IPackageData[]) => string;

/**
 * @alpha
 * Loose string type that represents the name of the generated license file.
 *
 * @example
 * ```ts
 * const licenseFileName: LicenseFileName = 'custom-license-file-name.html';
 * const licenseMarkdownFileName: LicenseFileName = 'custom-license-file-name.md';
 * const licenseTextFileName: LicenseFileName = 'custom-license-file-name.txt';
 * ```
 */

> export declare type LicenseFileName = `${string}.${'html' | 'md' | 'txt'}`; -->

Reviewer: ?

Author: Whoops! Will remove!
